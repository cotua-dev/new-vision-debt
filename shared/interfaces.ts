export interface PageProps {
    props: {};
    revalidate: number;
};
