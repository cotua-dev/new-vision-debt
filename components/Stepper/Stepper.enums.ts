export enum UserPurpose {
    ReduceYourMonthlyPayments = 80,
    ConsolidateYourPayments = 81,
    ReduceYourInterestRates = 82,
    AllOfTheAbove = 83,
};

export enum DebtType {
    CreditCards = 84,
    PersonalLoans = 85,
    MedicalBills = 86,
    AllOfTheAbove = 87,
};

export enum BehindPaymentsType {
    YesMore60 = 88,
    YesMore30 = 89,
    No = 90,
};

export enum FallBehindReason {
    LossOfIncome = 91,
    IncreasedCostOfLiving = 92,
    UnexpectedExpenses = 93,
    MedicalHardship = 94,
};

export enum Questions {
    UserPurpose = 'What are you looking to do?',
    DebtType = 'What kind of debts do you have?',
    BehindPaymentsType = 'Are you behind on your payments?',
    FallBehindReason = 'What made you fall behind?',
    MonthlyIncomeAmount = 'What is your monthly income?',
    UnsecuredDebtAmount = 'What is the total amount of your unsecured debt?',
    ZipCode = 'What city are you in?',
    Name = 'What is your name?',
    Email = 'What is your email?',
    Phone = 'Please provide your phone number for verification',
    Code = 'Please provide the code sent to your phone number',
};
