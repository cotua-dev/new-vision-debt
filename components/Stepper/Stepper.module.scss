@use 'styles/variables';
@use 'styles/mixins';

section.debt-advisors-stepper {
    @include mixins.base-section;
    @include mixins.flexify(center, center, column);
    height: 550px;
    padding: 10px;

    @include mixins.mobile {
        height: 100%;
        padding: 0 10px 50px;
    }

    div.steps-wrapper {
        div.step {
            text-align: center;
            margin: 50px 10px;

            @include mixins.custom(767px) {
                margin: 20px 10px;
            }

            &.hidden {
                display: none;
            }

            strong.question {
                @include mixins.main-heading;

                @include mixins.mobile {
                    font-size: 24px;
                }
            }
        }
    }

    small.error {
        font-size: 14px;
        font-weight: 700;
        color: variables.$secondary-color;
    }

    div.stepper-controls-wrapper {
        margin-top: 20px;
        display: flex;
        flex-direction: row;

        @include mixins.mobile {
            width: 100%;
        }

        button.previous-button,
        button.next-button {
            cursor: pointer;
            border: 0;
            border-radius: 10px;
            padding: 10px 0;
            color: #FFFFFF;
            font-family: 'Montserrat', Arial, Helvetica, sans-serif;
            font-size: 18px;
            font-weight: 300;
            @include mixins.flexify(center, center, row);

            &:disabled {
                cursor: not-allowed;
            }

            .caret {
                height: 20px;
                width: 20px;
            }
        }

        button.previous-button {
            width: 80px;
            background-color: variables.$secondary-color;

            &:disabled {
                background-color: lighten(variables.$secondary-color, 20%);
            }
        }

        button.next-button {
            width: 300px;
            background-color: variables.$primary-color;
            margin-left: 20px;

            &:disabled {
                background-color: lighten(variables.$primary-color, 20%);
            }

            @include mixins.mobile {
                width: 100%;
            }

            .caret {
                margin-left: 10px;
            }
        }
    }
}
