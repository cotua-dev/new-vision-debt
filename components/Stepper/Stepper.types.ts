export type MultipleChoiceValues = 0 | 1 | 2 | 3;

export type StepTypes = 'multiple-choice' | 'currency' | 'location' | 'phone' | 'verify' | 'name' | 'email';

export type StepperType = 'full' | 'short';
