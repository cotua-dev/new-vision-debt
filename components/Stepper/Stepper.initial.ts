import { StepperModel, Step } from './Stepper.interfaces';
import { Questions } from './Stepper.enums';

export const initialStepperModel: StepperModel = {
    userPurpose: null,
    debtType: null,
    behindPaymentsType: null,
    fallBehindReason: null,
    monthlyIncomeAmount: '',
    unsecuredDebtAmount: '',
    zipCode: '',
    firstName: '',
    lastName: '',
    email: '',
    phone: '',
    code: '',
    isMailer: false,
    sourceSite: '',
};

export const startingSteps: Step[] = [
    {
        question: Questions.UserPurpose,
        validity: false,
        property: ['userPurpose'],
        stepType: 'multiple-choice',
        choices: [
            { label: 'Reduce your monthly payments', value: 0 },
            { label: 'Consolidate your payments', value: 1 },
            { label: 'Reduce your interest rates', value: 2 },
            { label: 'All of the above', value: 3 },
        ],
        value: {},
    },
    {
        question: Questions.DebtType,
        validity: false,
        property: ['debtType'],
        stepType: 'multiple-choice',
        choices: [
            { label: 'Credit cards', value: 0 },
            { label: 'Personal loans', value: 1 },
            { label: 'Medical bills', value: 2 },
            { label: 'All of the above', value: 3 },
        ],
        value: {},
    },
    {
        question: Questions.BehindPaymentsType,
        validity: false,
        property: ['behindPaymentsType'],
        stepType: 'multiple-choice',
        choices: [
            { label: 'Yes - more than 60 days', value: 0 },
            { label: 'Yes - more than 30 days', value: 1 },
            { label: 'No', value: 2 },
        ],
        value: {},
    },
];

export const fallBehindReasonStep: Step[] = [
    {
        question: Questions.FallBehindReason,
        validity: false,
        property: ['fallBehindReason'],
        stepType: 'multiple-choice',
        choices: [
            { label: 'Loss of income', value: 0 },
            { label: 'Increase in cost of living', value: 1 },
            { label: 'Unexpected expenses', value: 2 },
            { label: 'Medical hardship', value: 3 },
        ],
        value: {},
    },
];

export const userSteps: Step[] = [
    {
        question: Questions.ZipCode,
        validity: false,
        property: ['zipCode'],
        stepType: 'location',
        value: {},
    },
    {
        question: Questions.Name,
        validity: false,
        property: ['firstName', 'lastName'],
        stepType: 'name',
        value: {},
    },
    {
        question: Questions.Email,
        validity: false,
        property: ['email'],
        stepType: 'email',
        value: {},
    },
    {
        question: Questions.Phone,
        validity: false,
        property: ['phone'],
        stepType: 'phone',
        value: {},
    },
    {
        question: Questions.Code,
        validity: false,
        property: ['code'],
        stepType: 'verify',
        value: {},
    },
];

export const endingSteps: Step[] = [
    {
        question: Questions.MonthlyIncomeAmount,
        validity: false,
        property: ['monthlyIncomeAmount'],
        stepType: 'currency',
        value: {},
    },
    {
        question: Questions.UnsecuredDebtAmount,
        validity: false,
        property: ['unsecuredDebtAmount'],
        stepType: 'currency',
        value: {},
    },
];

export const initialSteps: Step[] = [
    ...startingSteps,
    ...endingSteps,
    ...userSteps,
];

export const initialNoUserSteps: Step[] = [
    ...startingSteps,
    ...endingSteps,
];
