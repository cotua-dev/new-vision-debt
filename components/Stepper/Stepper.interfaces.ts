import { MultipleChoiceValues, StepperType, StepTypes } from './Stepper.types';
import { UserPurpose, DebtType, BehindPaymentsType, FallBehindReason, Questions } from './Stepper.enums';
import { MailerCheckResponseBody } from '../MailerCheck/MailerCheck.interfaces';

export interface MultipleChoice {
    label: string;
    value: MultipleChoiceValues;
};

export interface MailerModel {
    referenceId: string;
    middleInitial: string;
    suffix: string;
    address: string;
    city: string;
    stateAbbreviation: string;
    crrt: string;
    barcode: string;
    county: string;
    estimatedDebt: number;
    settled: number;
    newPayment: number;
    tier: string;
    segment: string;
    drop: string;
    tollFreeNumber: string;
    spanishTollFreeNumber: string;
    noticeDate: string;
    url: string;
    stateFull: string;
};

export interface StepperModel extends Partial<MailerModel> {
    userPurpose: MultipleChoiceValues | null;
    debtType: MultipleChoiceValues | null;
    behindPaymentsType: MultipleChoiceValues | null;
    fallBehindReason: MultipleChoiceValues | null;
    monthlyIncomeAmount: string;
    unsecuredDebtAmount: string;
    zipCode: string;
    firstName: string;
    lastName: string;
    email: string;
    phone: string;
    code: string;
    isMailer: boolean;
};

export interface ParsedStepperModel extends Partial<MailerModel> {
    userPurpose: UserPurpose;
    debtType: DebtType;
    behindPaymentsType: BehindPaymentsType;
    fallBehindReason: FallBehindReason;
    monthlyIncomeAmount: number;
    unsecuredDebtAmount: number;
    zipCode: string;
    firstName: string;
    lastName: string;
    email: string;
    phone: string;
    site: string;
    isMailer: boolean;
};

export interface Step {
    question: Questions;
    validity: boolean;
    property: (keyof StepperModel)[];
    stepType: StepTypes;
    choices?: MultipleChoice[];
    value: { [StepperModelKey in keyof StepperModel]?: null | string | number; };
};

export interface StepperProps {
    'stepper-type': StepperType,
    'user-info'?: MailerCheckResponseBody,
};
