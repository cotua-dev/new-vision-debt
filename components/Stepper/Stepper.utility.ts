import { UserPurpose, DebtType, BehindPaymentsType, FallBehindReason, Questions } from './Stepper.enums';
import { ParsedStepperModel, StepperModel } from './Stepper.interfaces';
import { MultipleChoiceValues } from './Stepper.types';

/**
 * Parse the user purpose multiple-choice field value
 * @param selection Number between 0 - 3 inclusive
 * @returns A number depending on the selection
 */
export function parseUserPurpose(selection: MultipleChoiceValues | null): UserPurpose {
    switch(selection) {
        case 0:
            return UserPurpose.ReduceYourMonthlyPayments;
        case 1:
            return UserPurpose.ConsolidateYourPayments;
        case 2:
            return UserPurpose.ReduceYourInterestRates;
        case 3:
            return UserPurpose.AllOfTheAbove;
        default:
            return UserPurpose.ReduceYourMonthlyPayments;
    }
}

/**
 * Parse the debt type multiple-choice field value
 * @param selection Number between 0 - 3 inclusive
 * @returns A number depending on the selection
 */
export function parseDebtType(selection: MultipleChoiceValues | null): DebtType {
    switch(selection) {
        case 0:
            return DebtType.CreditCards;
        case 1:
            return DebtType.PersonalLoans;
        case 2:
            return DebtType.MedicalBills;
        case 3:
            return DebtType.AllOfTheAbove;
        default:
            return DebtType.CreditCards;
    }
}

/**
 * Parse the behind payments type multiple-choice field value
 * @param selection Number between 0 - 3 inclusive
 * @returns A number depending on the selection
 */
export function parseBehindPaymentsType(selection: MultipleChoiceValues | null): BehindPaymentsType {
    switch(selection) {
        case 0:
            return BehindPaymentsType.YesMore60;
        case 1:
            return BehindPaymentsType.YesMore30;
        case 2:
            return BehindPaymentsType.No;
        default:
            return BehindPaymentsType.YesMore60;
    }
}

/**
 * Parse the fall behind reason multiple-choice field value
 * @param selection Number between 0 - 3 inclusive
 * @returns A number depending on the selection
 */
export function parseFallBehindReason(selection: MultipleChoiceValues | null): FallBehindReason {
    switch(selection) {
        case 0:
            return FallBehindReason.LossOfIncome;
        case 1:
            return FallBehindReason.IncreasedCostOfLiving;
        case 2:
            return FallBehindReason.UnexpectedExpenses;
        case 3:
            return FallBehindReason.MedicalHardship;
        default:
            return FallBehindReason.LossOfIncome;
    }
}

/**
 * Remove the dollar sign and commas from currency value
 * @param unparsedCurrencyValue String containing a dollar amount (Ex: "$10,000,000.00")
 * @returns Number minus commas and dollar sign (Ex: "10000000.00")
 */
export function parseCurrencyValue(unparsedCurrencyValue: string): number {
    return Number((unparsedCurrencyValue.replace(',', '')).replace('$', ''));
}

/**
 * Parse the phone number so it is acceptable for submission
 * @param unparsedPhoneNumber String containing a phone number (Ex: "+1 234 567 8901")
 * @returns Parsed string that is acceptable for submission (Ex: "+12345678901")
 */
export function parsePhoneNumber(unparsedPhoneNumber: string): string {
    return unparsedPhoneNumber
        .split('')
        .filter((letter: string) => (letter !== ' '))
        .join('');
}

/**
 * Parse the model so its values are acceptable for submission
 * @param unparsedModel Object containing values not acceptable for submission
 * @returns Object containing equivalent acceptable values
 */
export function parseModel(unparsedModel: StepperModel): ParsedStepperModel {
    return {
        userPurpose: parseUserPurpose(unparsedModel.userPurpose),
        debtType: parseDebtType(unparsedModel.debtType),
        behindPaymentsType: parseBehindPaymentsType(unparsedModel.behindPaymentsType),
        fallBehindReason: parseFallBehindReason(unparsedModel.fallBehindReason),
        monthlyIncomeAmount: parseCurrencyValue(unparsedModel.monthlyIncomeAmount),
        unsecuredDebtAmount: parseCurrencyValue(unparsedModel.unsecuredDebtAmount),
        zipCode: unparsedModel.zipCode,
        firstName: unparsedModel.firstName,
        lastName: unparsedModel.lastName,
        email: unparsedModel.email,
        phone: parsePhoneNumber(unparsedModel.phone),
        site: window.location.href,
        isMailer: unparsedModel.isMailer,
    };
}

/**
 * Send a request containing a phone number to receive a code via SMS for verification
 * @param phone String containing a phone number
 * @returns Promise containing the response object or undefined
 */
export async function sendSMS(phone: string): Promise<Response | undefined> {
    try {
        // Make sure `phone` is not just an empty string
        if (phone !== '') {
            // Remove spaces from `phone` to make it E.164 compliant
            const e164PhoneNumber: string = parsePhoneNumber(phone);

            // Send a request with the E.164 phone number that will send an SMS to the provided phone number
            // with a 6 digit code for verification
            const response = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/SendSMS`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ phoneNumber: e164PhoneNumber }),
            });

            // Return the response
            return response;
        }
    } catch(sendSMSError: unknown) {
        console.error({ sendSMSError });
        throw new Error(sendSMSError as string);
    }
}

/**
 * Send a request containing the phone number and code for verification
 * @param phone String containing a phone number
 * @param code String containing a code
 * @returns Promise containing the response object or undefined
 */
export async function verifySMSCode(phone: string, code: string): Promise<Response | undefined> {
    try {
        // Make sure `phone` and `code` are not just empty strings
        if (phone !== '' && code !== '') {
            // Remove spaces from `phone` to make it E.164 compliant
            const e164PhoneNumber: string = parsePhoneNumber(phone);

            // Send a request with the E.164 phone number and code for verification
            const response = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/VerifyCode`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ phoneNumber: e164PhoneNumber, code }),
            });

            // Return the response
            return response;
        }
    } catch(verifySMSCodeError: unknown) {
        console.error({ verifySMSCodeError });
        throw new Error(verifySMSCodeError as string);
    }
}

/**
 * Send request to add a new contact and deal in Bitrix
 * @param data Object containing already parsed data
 * @returns Promise containing the response object or undefined
 */
export async function addBitrixContactDeal(data: ParsedStepperModel): Promise<Response | undefined> {
    try {
        // Kick off request using the passed data object
        const response = await fetch(`${process.env.NEXT_PUBLIC_API_URL}/Bitrix`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(data),
        });

        // Return the response object
        return response;
    } catch(addBitrixContactDealError: unknown) {
        console.error({ addBitrixContactDealError });
        throw new Error(addBitrixContactDealError as string);
    }
}
