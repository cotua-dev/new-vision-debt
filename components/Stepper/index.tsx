import { useState } from 'react';
import { useRouter } from 'next/router';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faAngleRight, faAngleLeft, faCircleNotch } from '@fortawesome/free-solid-svg-icons';
import styles from './Stepper.module.scss';
import { ParsedStepperModel, Step, StepperModel, StepperProps } from './Stepper.interfaces';
import { initialNoUserSteps, initialStepperModel, initialSteps } from './Stepper.initial';
import { addBitrixContactDeal, parseCurrencyValue, parseModel, sendSMS, verifySMSCode } from './Stepper.utility';
import { Questions } from './Stepper.enums';
import { MultipleChoiceField } from '../fields/MultipleChoice';
import { Currency } from '../fields/Currency';
import { Location } from '../fields/Location';
import { Name } from '../fields/Name';
import { Email } from '../fields/Email';
import { Phone } from '../fields/Phone';
import { Verify } from '../fields/Verify';

export function Stepper(props: StepperProps): JSX.Element {
    const router = useRouter();
    const [model, setModel] = useState(initialStepperModel);
    const [currentStep, setCurrentStep] = useState(0);
    const [steps, setSteps] = useState(props['stepper-type'] == 'full' ? initialSteps : initialNoUserSteps);
    const [disableNextButton, setDisableNextButton] = useState(true);
    const [disablePreviousButton, setDisablePreviousButton] = useState(true);
    const [loading, setLoading] = useState(false);
    const [error, setError] = useState('');
    const [disableVerifyField, setDisableVerifyField] = useState(true);

    // Fields
    const [zipCode, setZipCode] = useState(props['user-info'] === undefined ? '' : props['user-info'].zip);
    const [firstName, setFirstName] = useState(props['user-info'] === undefined ? '' : props['user-info'].firstName);
    const [lastName, setLastName] = useState(props['user-info'] === undefined ? '' : props['user-info'].lastName);
    const [email, setEmail] = useState('');
    const [phone, setPhone] = useState('');
    const [code, setCode] = useState('');
    const [usState, setUSState] = useState(props['user-info'] === undefined ? '' : props['user-info'].stateAbbreviation);

    /**
     * Send a SMS to the provided phone number to verify the user owns the number
     */
    async function kickOffSMS(): Promise<void> {
        // Show a loading spinner, disable verify field, and reset error
        setLoading(true);
        setDisableVerifyField(true);
        setError('');

        // Make sure the phone is not an empty string
        if (phone !== '') {
            const smsResponse: Response | undefined = await sendSMS(phone);

            // Make sure we have a response object and its status is 200
            if (smsResponse !== undefined && smsResponse.status === 200) {
                // Enable verify field
                setDisableVerifyField(false);
            } else {
                // Display error
                setError('Something went wrong. Please try again');
            }
        } else {
            // Display error
            setError('Please provide a valid phone number');
        }

        // Hide the loading spinner
        setLoading(false);
    }

    function sendThankYouPage() {
        switch (router.pathname) {
            case '/ohio':
                window.location.href = `${window.location.origin}/ohio/thank-you`;
                break;
            case '/new-jersey':
                window.location.href = `${window.location.origin}/new-jersey/thank-you`;
                break;
            case '/colorado':
                window.location.href = `${window.location.origin}/colorado/thank-you`;
                break;
            case '/credit-card':
                window.location.href = `${window.location.origin}/credit-card/thank-you`;
                break;
            case '/debt':
                window.location.href = `${window.location.origin}/debt/thank-you`;
                break;
            case '/medical':
                window.location.href = `${window.location.origin}/medical/thank-you`;
                break;
            case '/personal':
                window.location.href = `${window.location.origin}/personal/thank-you`;
                break;
            case '/student-loan':
                window.location.href = `${window.location.origin}/student-loan/thank-you`;
                break;
            default:
                window.location.href = `${window.location.origin}/thank-you`;
                break;
        }
    }

    /**
     * Verify the user provided code with their phone number
     */
    async function submission(data: ParsedStepperModel): Promise<void> {
        // Show a loading spinner and reset error
        setLoading(true);
        setError('');

        if (props['stepper-type'] === 'short') {
            const bitrixResponse: Response | undefined = await addBitrixContactDeal(data);

            if (bitrixResponse !== undefined && bitrixResponse.status === 201) {
                sendThankYouPage();
            } else {
                setError('Something went wrong. Please try again');
            }

            // Stop the function here
            setLoading(false);
            return;
        }

        // Make sure we have a `phone` and `code`
        if (phone !== '' && code !== '' && props['stepper-type'] === 'full') {
            const verifyResponse: Response | undefined = await verifySMSCode(phone, code);

            // Make sure we have a response object and its status is 200
            if (verifyResponse !== undefined && verifyResponse.status === 200) {
                const bitrixResponse: Response | undefined = await addBitrixContactDeal(data);

                // Make sure we have a response object and its status is 201
                if (bitrixResponse !== undefined && bitrixResponse.status === 201) {
                    // Send to thank you page with browser refresh (this way state is completely wiped in one go)
                    sendThankYouPage();
                } else {
                    // Display error
                    setError('Something went wrong. Please try again');
                }
            } else {
                // Display error
                setError('Something went wrong. Please try again');
            }
        } else {
            // Display error
            setError('Please provide a valid phone number and code for verification');
        }

        // Hide the loading spinner
        setLoading(false);
    }

    /**
     * @todo Needs work
     */
    function previousStep() {
        if (currentStep - 1 < 0) {
            setDisablePreviousButton(true);
            setCurrentStep(0);
        } else {
            setDisablePreviousButton(false);
            setCurrentStep(currentStep - 1);
        }

        setDisableNextButton(false);
    }

    /**
     * @todo Needs work
     */
    function nextStep() {
        // Grab needed properties from the current step
        const { validity, question } = steps[currentStep];

        // Check if we passed the unsecured debt field
        if (question === Questions.UnsecuredDebtAmount) {
            // Parse the unsecured debt amount into a number
            const unsecuredDebtAmountNumber: number = parseCurrencyValue(model.unsecuredDebtAmount);

            // Check if the debt is less than $5,000
            if (unsecuredDebtAmountNumber < 5000) {
                // If so, send to disqualify page with browser refresh (this way state is completely wiped in one go)
                window.location.href = `${window.location.origin}/dq`;
            }
        }

        // Check if we passed the zip code field
        if (question === Questions.ZipCode) {
            // Check to see if the city is in North Dakota or South Carolina
            if (usState === 'North Dakota' || usState === 'South Carolina') {
                // If so, send to disqualify page with browser refresh (this way state is completely wiped in one go)
                window.location.href = `${window.location.origin}/dq`;
            }
        }

        // Check if we passed the phone number field
        if (question === Questions.Phone) {
            // We should have a phone number at this point which means we can send
            // a SMS message to verify the phone number
            kickOffSMS();
        }

        // Check the current step's validity
        if (validity) {
            // If valid, enable the next button
            setDisableNextButton(false);

            // Check if we are on the last step
            if (currentStep >= steps.length - 1) {
                // Disable the next button and set the current step to the last step
                // since we should not go any further
                setDisableNextButton(true);
                setCurrentStep(steps.length - 1);
            } else {
                // Otherwise, go to the next step
                setCurrentStep(currentStep + 1);
            }
        } else {
            // If currently invalid, disable the next button
            setDisableNextButton(true);
        }

        // Enable the previous button since we are moving forward
        setDisablePreviousButton(false);
    }

    /**
     * Handle the submit button click event
     */
    function onSubmit(): void {
        // Get the unparsed model and parse it for submission
        const unparsedModel: StepperModel = {
            ...model,
            email,
            firstName,
            lastName,
            phone,
            zipCode,
        };
        let parsedModel: ParsedStepperModel = parseModel(unparsedModel);

        if (props['stepper-type'] === 'short') {
            parsedModel.isMailer = true;
        }

        if (parsedModel.isMailer && props['user-info'] !== undefined) {
            const {
                referenceId, middleInitial, suffix, address, city,
                stateAbbreviation, crrt, barcode, county, estimatedDebt,
                settled, newPayment, tier, segment, drop, tollFreeNumber,
                spanishTollFreeNumber, noticeDate, url, stateFull,
            } = props['user-info'];

            parsedModel = {
                ...parsedModel,
                referenceId, middleInitial, suffix, address, city,
                stateAbbreviation, crrt, barcode, county, estimatedDebt,
                settled, newPayment, tier, segment, drop, tollFreeNumber,
                spanishTollFreeNumber, noticeDate, url, stateFull,
            };
        }

        // Run submission
        submission(parsedModel);
    }

    return (
        <section className={styles['debt-advisors-stepper']}>
            <small>{`${currentStep + 1} of ${steps.length}`}</small>
            <div className={styles['steps-wrapper']}>
                {steps.map((step: Step) => {
                    const stepIndex: number = steps.findIndex((thisStep: Step) => thisStep.question === step.question);
                    const showStep: boolean = stepIndex === currentStep;

                    return (
                        <div key={step.question} className={`${styles['step']} ${showStep ? '' : styles['hidden']}`}>
                            <strong className={styles['question']}>{step.question}</strong>
                            {
                                step.stepType === 'multiple-choice' &&
                                    <MultipleChoiceField
                                        choices={step.choices !== undefined ? step.choices : []}
                                        field={step.property}
                                        setModel={setModel}
                                        setSteps={setSteps}
                                        model={model}
                                        nextStep={nextStep}
                                        steps={steps}
                                        currentStep={currentStep}
                                        stepperType={props['stepper-type']}
                                    />
                            }
                            {
                                step.stepType === 'currency' &&
                                    <Currency
                                        field={step.property}
                                        setModel={setModel}
                                        model={model}
                                        steps={steps}
                                        currentStep={currentStep}
                                        setDisableNextButton={setDisableNextButton}
                                    />
                            }
                            {
                                step.stepType === 'location' &&
                                    <Location
                                        steps={steps}
                                        currentStep={currentStep}
                                        setDisableNextButton={setDisableNextButton}
                                        zipCode={zipCode}
                                        setZipCode={setZipCode}
                                        setUSState={setUSState}
                                    />
                            }
                            {
                                step.stepType === 'name' &&
                                    <Name
                                        steps={steps}
                                        currentStep={currentStep}
                                        setDisableNextButton={setDisableNextButton}
                                        firstName={firstName}
                                        lastName={lastName}
                                        setFirstName={setFirstName}
                                        setLastName={setLastName}
                                    />
                            }
                            {
                                step.stepType === 'email' &&
                                    <Email
                                        steps={steps}
                                        currentStep={currentStep}
                                        setDisableNextButton={setDisableNextButton}
                                        email={email}
                                        setEmail={setEmail}
                                    />
                            }
                            {
                                step.stepType === 'phone' &&
                                    <Phone
                                        steps={steps}
                                        currentStep={currentStep}
                                        setDisableNextButton={setDisableNextButton}
                                        phone={phone}
                                        setPhone={setPhone}
                                    />
                            }
                            {
                                step.stepType === 'verify' &&
                                    <Verify
                                        steps={steps}
                                        currentStep={currentStep}
                                        setDisableNextButton={setDisableNextButton}
                                        code={code}
                                        setCode={setCode}
                                        disableVerifyField={disableVerifyField}
                                    />
                            }
                        </div>
                    );
                })}
            </div>
            {loading && <FontAwesomeIcon className="h-7 text-daa-purple animate-spin" icon={faCircleNotch} spin/>}
            {error !== '' && <small className={styles['error']}>{error}</small>}
            <div className={styles['stepper-controls-wrapper']}>
                <button
                    className={styles['previous-button']}
                    type="button"
                    onClick={() => previousStep()}
                    disabled={disablePreviousButton}
                ><FontAwesomeIcon className={styles['caret']} icon={faAngleLeft}/></button>
                {
                    steps.length - 1 !== currentStep &&
                    <button
                        className={styles['next-button']}
                        type="button"
                        onClick={() => nextStep()}
                        disabled={disableNextButton}
                    >{'Next'}<FontAwesomeIcon className={styles['caret']} icon={faAngleRight}/></button>
                }
                {
                    steps.length - 1 === currentStep &&
                    <button
                        className={styles['next-button']}
                        type="button"
                        onClick={() => onSubmit()}
                        disabled={disableNextButton}
                    >{`Call or Text Me Now`}</button>
                }
            </div>
        </section>
    );
}
