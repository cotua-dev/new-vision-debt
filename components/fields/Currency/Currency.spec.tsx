import { render, RenderResult } from '@testing-library/react';
import { SharedFieldProps } from '../shared.interfaces';
import { initialStepperModel } from '../../Stepper/Stepper.initial';
import { Currency } from './index';

describe('Currency Component', (): void => {
    let renderResult: RenderResult;
    const props: SharedFieldProps = {
        field: ['behindPaymentsType'],
        setModel: () => {},
        model: initialStepperModel,
        steps: [],
        currentStep: 0,
    };

    beforeAll((): RenderResult => renderResult = render(<Currency {...props}/>));

    test('renders', (): void => {
        const { container } = renderResult;
        expect(container.firstChild).toMatchSnapshot();
    });

    afterAll((): void => renderResult.unmount());
});
