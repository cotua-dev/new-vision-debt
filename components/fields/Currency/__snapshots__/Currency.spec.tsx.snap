// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency Component renders 1`] = `
<div
  class="currency-field-wrapper"
>
  <input
    class="currency-field"
    inputmode="numeric"
    type="text"
    value=""
  />
</div>
`;
